### PR TITLE
Fix reading past comments after an end object or end array token within Utf8JsonReader

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/40385

This fix needs to be ported to 3.0

cc @IonKiwi, @am11, @bartonjs, @stephentoub, @steveharter, @ericstj

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 1]
```diff
@@ -1899,7 +1887,37 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentTokenMultiSegment()
             }
             else
             {
-                goto RollBack;
+                Debug.Assert(_tokenType == JsonTokenType.EndArray || _tokenType == JsonTokenType.EndObject);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: This is the fix for the underlying issue. We weren't dealing with the "end" token types.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 2]
```diff
@@ -1846,48 +1846,36 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentTokenMultiSegment()
...
+                goto Done;
             }
             else if (_tokenType == JsonTokenType.StartArray)
             {
-                if (first == JsonConstants.CloseBracket)
-                {
-                    EndArray();
-                }
-                else
+                Debug.Assert(first != JsonConstants.CloseBracket);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Replacing unreachable branch with Debug.Assert (`first == JsonConstants.CloseBracket` is covered above) .

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 3]
```diff
@@ -1846,48 +1846,36 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentTokenMultiSegment()
             }
             else if (_tokenType == JsonTokenType.StartObject)
             {
-                if (first == JsonConstants.CloseBrace)
+                Debug.Assert(first != JsonConstants.CloseBrace);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Replacing unreachable branch with Debug.Assert (`first == JsonConstants.CloseBrace` is covered above) .

**Comment by ahsonkhan** [comment]
Unrelated `System.Security.Cryptography.Csp.Tests` test failure on corefx-ci (Windows Build x86_Release):
https://dev.azure.com/dnceng/public/_build/results?buildId=316342
```text
 System.Security.Cryptography.Rsa.Tests.EncryptDecrypt_Span.Decrypt_WrongKey_Pkcs7 [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.Security.Cryptography.CryptographicException)
      Actual:   (No exception was thrown)
      Stack Trace:
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/EncryptDecrypt.netcoreapp.cs(120,0): at System.Security.Cryptography.Rsa.Tests.EncryptDecrypt_Span.Decrypt_WrongKey(RSAEncryptionPadding padding)
        /_/src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/EncryptDecrypt.netcoreapp.cs(94,0): at System.Security.Cryptography.Rsa.Tests.EncryptDecrypt_Span.Decrypt_WrongKey_Pkcs7()
```

Looks like its a known issue: https://github.com/dotnet/corefx/pull/38172#issuecomment-498328482
cc @bartonjs

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 4]
```diff
@@ -1846,48 +1846,36 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentTokenMultiSegment()
             }
             else if (_tokenType == JsonTokenType.StartObject)
             {
-                if (first == JsonConstants.CloseBrace)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: Per offline discussion this was this unreachable to begin with (so not part of the issue)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 5]
```diff
@@ -1980,46 +1980,34 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentToken()
...
+                    _tokenType = JsonTokenType.StartObject;
+                    _bytePositionInLine = prevPosition;
+                    _lineNumber = prevLineNumber;
+                    goto RollBack;
                 }
+                goto Done;
             }
             else if (_tokenType == JsonTokenType.StartArray)
             {
-                if (first == JsonConstants.CloseBracket)
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: Per offline discussion this was this unreachable to begin with (so not part of the issue)

File: src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 6]
```diff
@@ -1728,6 +1728,284 @@ private static void TrailingCommasHelper(ReadOnlySequence<byte> utf8, JsonReader
...
+        [InlineData("{\"Property1\": {\"Property1.1\": 42}, /*comment*/\"Property2\": 42}")]
+        [InlineData("{\"Property1\": {\"Property1.1\": 42} // comment\n,\"Property2\": 42}")]
+        [InlineData("{\"Property1\": {\"Property1.1\": 42}, // comment\n // comment\n\"Property2\": 42}")]
+        [InlineData("[{\"Property1\": 42}, // comment\n 42]")]
+        [InlineData("[{\"Property1\": 42} // comment\n, 42]")]
+        [InlineData("[{\"Property1\": 42, // comment\n \"prop\": 43 }, // comment\n 42]")]
+        [InlineData("[{\"Property1\": \"42\", // comment\n \"prop\": 43 }, // comment\n 42]")]
+        [InlineData("[{\"Property1\": true, // comment\n \"prop\": 43 }, // comment\n 42]")]
+        [InlineData("[{\"Property1\": false, // comment\n \"prop\": 43 }, // comment\n 42]")]
+        [InlineData("[{\"Property1\": null, // comment\n \"prop\": 43 }, // comment\n 42]")]
```
Comment (steveharter) [inline src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: Would it be redundant to have a nested object and\or nested array InlineData? To ensure the changes handle multiple levels.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs [hunk 7]
```diff
@@ -1899,7 +1887,37 @@ private ConsumeTokenResult ConsumeNextTokenFromLastNonCommentTokenMultiSegment()
             }
             else
             {
-                goto RollBack;
+                Debug.Assert(_tokenType == JsonTokenType.EndArray || _tokenType == JsonTokenType.EndObject);
+                if (_inObject)
+                {
+                    Debug.Assert(first != JsonConstants.CloseBracket);
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs]: I notice there weren't (many) Asserts like this previously, but they seem useful. Perhaps add them in the future to other areas?

File: src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 8]
```diff
@@ -1728,6 +1728,284 @@ private static void TrailingCommasHelper(ReadOnlySequence<byte> utf8, JsonReader
...
+        [InlineData("{\"Property1\": {\"Property1.1\": 42} // comment\n,}")]
+        [InlineData("{\"Property1\": {\"Property1.1\": 42}, // comment\n // comment\n}")]
+        [InlineData("[{\"Property1\": 42}, // comment\n ]")]
+        [InlineData("[{\"Property1\": 42} // comment\n, ]")]
+        [InlineData("[{\"Property1\": 42, // comment\n }, // comment\n ]")]
+        [InlineData("[{\"Property1\": \"42\", // comment\n }, // comment\n ]")]
+        [InlineData("[{\"Property1\": true, // comment\n }, // comment\n ]")]
+        [InlineData("[{\"Property1\": false, // comment\n }, // comment\n ]")]
+        [InlineData("[{\"Property1\": null, // comment\n }, // comment\n ]")]
+        public static void ReadJsonStringsWithCommentsAndTrailingCommas(string jsonString)
```
Comment (steveharter) [inline src/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs]: Was there an issue with trailing commas?

**Pull request merged**